mmv1/third_party/terraform/services/artifactregistry/data_source_artifact_registry_docker_image.go

```diff
@@ -0,0 +1,331 @@
+package artifactregistry
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-provider-google/google/tpgresource"
+	transport_tpg "github.com/hashicorp/terraform-provider-google/google/transport"
+)
+
+// https://cloud.google.com/artifact-registry/docs/reference/rest/v1/projects.locations.repositories.dockerImages#DockerImage
+type DockerImage struct {
+	name           string
+	uri            string
+	tags           []string
+	imageSizeBytes string
+	mediaType      string
+	uploadTime     string
+	buildTime      string
+	updateTime     string
+}
+
+func DataSourceArtifactRegistryDockerImage() *schema.Resource {
+
+	return &schema.Resource{
+		Read: DataSourceArtifactRegistryDockerImageRead,
+
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: `Project ID of the project.`,
+			},
+			"repository": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `The repository name.`,
+			},
+			"region": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `The location of the artifact registry repository. For example, "us-west1".`,
+			},
+			"image": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `The image name to fetch.`,
+			},
+			"digest": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: `The image digest to fetch.  This cannot be used if tag is provided.`,
+			},
+			"tag": {
```

Comment:
Regarding the field names, this one in particular is a bit ambiguous, because the resource already has a list of `tags`, and this is really more like a query/filter parameter. You may be able to find precedence for something like this as well, but if not I'll try to look around for similar data source patterns in the provider.

Reply:
Yea, `tag` and `tags` would easily be mixed up. The intent of the issue was to have a data source that would return the URI of a particular `image:tag` docker image, so I think it's fine to keep those two field names.  I added the `tags` simply because it was available information from the REST API -- it wasn't specifically asked for in the issue, but I could see someone using the tags information when populating the `description` field for a compute instance or cloud run service.

I'd think renaming the `tags` field to something less ambiguous would be good, maybe `all_image_tags` or `all_tags`?

Reply:
@roaks3 I found that the [google_compute_regions](https://registry.terraform.io/providers/hashicorp/google/latest/docs/data-sources/compute_regions) uses the `status` field as a filter parameter. Could this be sufficient precedence for the `tag` field used in this data source?

Reply:
Yea, I think that makes sense in terms of some precedence. I don't see this come up very much in existing datasources, so the `tag` + `tags` might be the best we can do in terms of modeling.

However, looking at https://registry.terraform.io/providers/calxus/docker/latest/docs/data-sources/image, we may want to consider using just `image_name`, and having the tag and digest inferred from that value. This would be particularly appealing if we don't need to parse anything client-side, and can send the name up to the server for each case, because then we can rely on the server logic and the client has less ways to break or surprise users. Based on your current implementation though, I suspect the server doesn't recognize a tag in the image name?

Reply:
@roaks3 Sounds good!  I've implemented the logic to infer tags and digest from `image_name`, and removed those arguments from the data source.